x-pack/legacy/plugins/apm/server/lib/services/get_services/index.ts

```diff
@@ -6,26 +6,25 @@
 
 import { isEmpty } from 'lodash';
 import { PromiseReturnType } from '../../../../typings/common';
-import { Setup } from '../../helpers/setup_request';
+import { SetupWithAllFilters } from '../../helpers/setup_request';
 import { hasHistoricalAgentData } from './has_historical_agent_data';
 import { getLegacyDataStatus } from './get_legacy_data_status';
 import { getServicesItems } from './get_services_items';
 
 export type ServiceListAPIResponse = PromiseReturnType<typeof getServices>;
-export async function getServices(setup: Setup) {
-  const items = await getServicesItems(setup);
-  const hasLegacyData = await getLegacyDataStatus(setup);
 
-  // conditionally check for historical data if no services were found in the current time range
+export async function getServices(setup: SetupWithAllFilters) {
+  const itemsPromise = getServicesItems(setup);
+  const hasLegacyDataPromise = getLegacyDataStatus(setup);
+  const items = await itemsPromise;
```

Comment:
Any reason for not `await`'ing immediately? This ends up being `await`'ed in serial endyway.

Reply:
`getLegacyDataStatus` (1) is not blocked by `getServicesItems` (2), only `hasHistoricalAgentData` is. So this starts 1 and 2 immediately, and if 1 finishes, 3 starts. Then the function returns when all promises are resolved. At least, that's what it _should_ do. Do you see something else happening?

I could also rewrite with promises if it makes more sense:

```ts
const [{ items, hasHistoricalData }, hasLegacyData] = await Promise.all([
  getServicesItems(setup).then(async servicesItems => {
    const noDataInCurrentTimeRange = isEmpty(servicesItems);

    return {
      items: servicesItems,
      hasHistoricalData: noDataInCurrentTimeRange
        ? await hasHistoricalAgentData(setup)
        : true
    };
  }),
  getLegacyDataStatus(setup)
]);
```

WDYT?

Reply:
Maybe I'm missing something but isn't this equivalent to using `Promise.all` for the first two promises?

```ts
export async function getServices(setup: SetupWithAllFilters) {
  const [items, hasLegacyData] = await Promise.all([
    getServicesItems(setup),
    getLegacyDataStatus(setup)
  ]);

  const noDataInCurrentTimeRange = isEmpty(items);
  const hasHistoricalData = noDataInCurrentTimeRange
    ? await hasHistoricalAgentData(setup)
    : true;

  return {
    items,
    hasHistoricalData,
    hasLegacyData
  };
}
```

Reply:
At least this approach is a lot more clear to me.

Reply:
`getServicesItems` could finish before `getLegacyDataStatus` and then unnecessary block `hasHistoricalAgentData`, right? (`hasHistoricalAgentData` needs `getServicesItems`, but not `hasLegacyData`)

Reply:
Maybe it's not worth it though? I do think the version that we have today is more readable for sure.

Reply:
> `getServicesItems` could finish before `getLegacyDataStatus` and then the latter would unnecessary block `hasHistoricalAgentData`, right?

Did you time this? I'd expect in most cases `getLegacyDataStatus` to be faster than `getServicesItems` - ofc this depends on both the selected time range and the data volumes.
Regardless, I think the difference is negligible and I'd optimize for the most readable version.

Reply:
I didn't time it, I figured it would be a win anyway but agreed, let's keep it simple 👍